### PR TITLE
Add gurkancount command

### PR DESCRIPTION
**Pull request by gustavwilliam**
## Description

This command loops through the display names of all members and checks which ones contain "gurkan".

### Important note:
This enables member intents, which means that it'll have to be enabled on [Discord Developers Portal](https://discord.com/developers/applications), if it isn't already. It also means that we'll be tracking member intents.

This is needed, since the display names of guild members will be checked.

## Screenshots

Here is the new command:
Screenshot 2020-11-15 at 04 23 41

## Did you:

- [x] Join the [**Python Discord Community**](https://discord.gg/python)?
- [x] If dependencies have been added or updated, run `pipenv lock`?
- [x] **Lint your code** (`pipenv run lint`)?
- [x] Set the PR to **allow edits from contributors**?

### CODE REVIEW

File: bot/exts/evergreen/gurkan.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+
+class Gurkan(commands.Cog):
+    """Commands related to Gurkan and the Gurkult."""
+
+    def __init__(self, bot: commands.Bot):
+        self.bot = bot
+
+    @commands.command(name="gurkancount", aliases=("gc", "gurkcount"))
+    async def gurkancount(self, ctx: commands.Context) -> None:
```
Comment (fisher60) [inline bot/exts/evergreen/gurkan.py:14]: This function name could be clearer if it was using `snake_case` i.e `gurkan_count`

File: bot/exts/evergreen/gurkan.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        embed = discord.Embed(
+            title="Gurkancount",
+            color=Colours.bright_green,
+            description=f"There are a total of `{gurkan_count}` gurkans in the server!")
+
+        await ctx.send(embed=embed)
+
+
+def setup(bot: commands.Bot) -> None:
+    """Load the Bookmark cog."""
```
Comment (fisher60) [inline bot/exts/evergreen/gurkan.py:28]: Forgot to change the docstring to the `Gurkan Cog` here.

**Review by fisher60 — CHANGES_REQUESTED** [review]
There is not much to review here. Mostly style preferences. Only really need to change the docstring.

The code ran perfectly fine, the only real concern would be the performance of looping through so many members.

File: bot/exts/evergreen/gurkan.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    @commands.command(name="gurkancount", aliases=("gc", "gurkcount"))
+    async def gurkancount(self, ctx: commands.Context) -> None:
+        """Get the amount of members with 'gurkan' in their names."""
+        gurkans = [1 for member in ctx.guild.members if "gurkan" in member.display_name.lower()]
+        gurkan_count = sum(gurkans)
+
+        embed = discord.Embed(
+            title="Gurkancount",
+            color=Colours.bright_green,
+            description=f"There are a total of `{gurkan_count}` gurkans in the server!")
```
Comment (Den4200) [inline bot/exts/evergreen/gurkan.py:22]: The closing parenthesis should be on the next line.

File: bot/exts/evergreen/gurkan.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    """Commands related to Gurkan and the Gurkult."""
+
+    def __init__(self, bot: commands.Bot):
+        self.bot = bot
+
+    @commands.command(name="gurkancount", aliases=("gc", "gurkcount"))
+    async def gurkancount(self, ctx: commands.Context) -> None:
+        """Get the amount of members with 'gurkan' in their names."""
+        gurkans = [1 for member in ctx.guild.members if "gurkan" in member.display_name.lower()]
+        gurkan_count = sum(gurkans)
```
Comment (Den4200) [inline bot/exts/evergreen/gurkan.py:17]: Maybe something like this instead? I'm not sure how I feel about having two variables for this.
```suggestion
        gurkan_count = sum(
            1 for member in ctx.guild.members if "gurkan" in member.display_name.lower()
        )
```

**Review by Den4200 — CHANGES_REQUESTED** [review]
There are all purely stylistic changes!

File: bot/exts/evergreen/gurkan.py (hunk 1 above)
Comment (gustavwilliam) [reply bot/exts/evergreen/gurkan.py:14]: I'm not sure about this one, since it's a command. If it was a usual method, it would be using snake_case, but this is a command.

**Comment by Den4200** [comment]
> There is not much to review here. Mostly style preferences. Only really need to change the docstring.
> 
> The code ran perfectly fine, the only real concern would be the performance of looping through so many members.

@fisher60 It takes very, very little resources loop through the members, so no worries here. It's only about 100,000 iterations for our server size.

File: bot/exts/evergreen/gurkan.py (hunk 1 above)
Comment (Den4200) [reply bot/exts/evergreen/gurkan.py:14]: What's wrong with having the name as `snake_case`?

File: bot/exts/evergreen/gurkan.py (hunk 4 above)
Comment (gustavwilliam) [reply bot/exts/evergreen/gurkan.py:17]: Looks much better. I didn't quite like having all of that on a single line nor creating two variables for it, so this solution looks better in some sense.

The only real reason that having two variables would be worth considering is if we would like to add some feature to do additional filtering in the future. That can be added when it becomes relevant, though, so this change looks good to me.

File: bot/exts/evergreen/gurkan.py (hunk 1 above)
Comment (gustavwilliam) [reply bot/exts/evergreen/gurkan.py:14]: We definitely could. I didn't think that was the convention for commands, though. Let's do it.

Comment (Den4200) [reply bot/exts/evergreen/gurkan.py:14]: The command name is `gurkancount`, as you set it in the decorator, but the method name should be `snake_case` instead.

**Review by Den4200 — APPROVED** [review]
Looks good to me.

**Review by lemonsaurus — CHANGES_REQUESTED** [review]
Let's make it more generic. I'd like to know how many users fit into _any_ arbitrary nickname feature.  We used to have `.lemoncount`, and I think there are still dozens. Let's have a feature that will count number of users with nicknames that contain any substring?

**Comment by gustavwilliam** [comment]
@lemonsaurus We could do that. The question is if we really want users to be able to search for any terms in display names or only a set of terms. What do you think about this?

**Comment by lemonsaurus** [comment]
what would be really cool is if you could hijack any command that ended with _count_ , so that `.gurkancount` and `.lemoncount` and `.vescount` would all just work.

**Comment by gustavwilliam** [comment]
That would be really interesting. Sadly, it would interfere with the emojicount command. If we were to ever do that (which would be super cool), we'd have to turn emoji count into a command + subcommand or something similar.

Either way, if we keep in within the scope of this PR, I wouldn't mind adding an optional argument with the string to search for. I think keeping the command name `gurkancount` would also be slightly funny and work well.

Then again, do we want people to search for *any* term? I don't see any clear issues with that. We have great moderation, so inappropriate terms would be handlers just like with anything.

**Comment by Den4200** [comment]
> what would be really cool is if you could hijack any command that ended with _count_ , so that `.gurkancount` and `.lemoncount` and `.vescount` would all just work.

This is a cool idea, I like it.


> That would be really interesting. Sadly, it would interfere with the emojicount command. If we were to ever do that (which would be super cool), we'd have to turn emoji count into a command + subcommand or something similar.

This shouldn't be an issue, since the logic would have to be implemented in the error handler.


> Then again, do we want people to search for *any* term? I don't see any clear issues with that. We have great moderation, so inappropriate terms would be handlers just like with anything.

I don't see it as an issue either.

**Comment by gustavwilliam** [comment]
I think let's just start with implementing it where you have a second optional argument for specifying the name. If we want to expand it to work with every count, let's do that later.

**Comment by SebastiaanZ** [comment]
One thing to consider is that SeasonalBot currently does not have the "members intent", which means that it also does not receive those events from Discord. Enabling the intent does increase the resource usage of Seasonal Bot (although the "presences intent" is probably way more heavy). I think I'm just a bit confused as to why we'd want this count command and if it's worth the additional load it will have on our back-end.

**Comment by gustavwilliam** [comment]
@SebastiaanZ yes. That's an important question to ask. I don't know exactly how much that would cost to have, but is it something that we have resources for? While we shouldn't add commands just because we can, I also think it would be a fun command to have — which is an important part for SeasonalBot.

As for the reasoning behind having a command like this, I think there's an interest in counting the amount of members with "gurkan" in their display names. You have probably noticed that the count is quite significant. As @lemonsaurus mentions, it would also be fun to check other names, like "lemon" and "ves."

TL;DR: It's a fun command that I think many would appreciate.

**Comment by lemonsaurus** [comment]
hmm, ves makes a good point about user intents. actually the resource load difference would probably be a bit significant.

this seemed like a better idea while I was drunk.

**Comment by gustavwilliam** [comment]
Is that something we don't have resources for?

**Comment by kwzrd** [comment]
We do have resources for it, the question is more whether we want to use them this way. Before intents rolled out, every bot was subscribed by default. When I was working on #478, core dev consensus seemed to be that we'd rather not have the intent if feasible. I'm impartial, but I don't know if this is a feature that would justify adding the intent, if we don't otherwise want it. If it's well implemented, generic and would be a useful utility to e.g. mods, we can also consider adding it to the main bot, which does subsribe to the intent.

**Comment by lemonsaurus** [comment]
yeah, I'm afraid I think member intents is too high a price to pay for this feature. Sorry, gurk.
